jupiterone/resource_rule.go

```diff
@@ -4,342 +4,627 @@ import (
+	resp.TypeName = req.ProviderTypeName + "_rule"
+}
+
+// Configure implements resource.ResourceWithConfigure
+func (r *QuestionRuleResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	p, ok := req.ProviderData.(*JupiterOneProvider)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected JupiterOneProvider, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.version = p.version
+	r.client = p.Client
+}
+
+// Schema implements resource.ResourceWithConfigure
+func (*QuestionRuleResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
 	var RulePollingIntervals = []string{"DISABLED", "THIRTY_MINUTES", "ONE_HOUR", "ONE_DAY", "ONE_WEEK"}
 
-	return &schema.Resource{
-		CreateContext: createQuestionRuleInstanceResource,
-		ReadContext:   resourceQuestionRuleInstanceRead,
-		UpdateContext: resourceQuestionRuleInstanceUpdate,
-		DeleteContext: resourceQuestionRuleInstanceDelete,
-		Schema: map[string]*schema.Schema{
-			"name": {
-				Type:         schema.TypeString,
-				Required:     true,
-				Description:  "Name of the rule, which is unique to each account.",
-				ValidateFunc: validation.StringLenBetween(MIN_RULE_NAME_LENGTH, MAX_RULE_NAME_LENGTH),
+	resp.Schema = schema.Schema{
+		Description: "A saved JupiterOne question based alert",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Description: "Unique id that identifies the rule",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"version": schema.Int64Attribute{
+				Description: "Computed current version of the rule. Incremented each time the rule is updated.",
+				Computed:    true,
+				PlanModifiers: []planmodifier.Int64{
+					// This works with ModifyPlan() prevent planned changes
+					// to this computed value
+					int64planmodifier.UseStateForUnknown(),
+				},
 			},
-			"description": {
-				Type:        schema.TypeString,
+			"name": schema.StringAttribute{
+				Required:    true,
+				Description: "Name of the rule, which is unique to each account.",
+				Validators: []validator.String{
+					stringvalidator.LengthBetween(MIN_RULE_NAME_LENGTH, MAX_RULE_NAME_LENGTH),
+				},
+			},
+			"description": schema.StringAttribute{
 				Description: "Description of the rule",
 				Required:    true,
 			},
-			"spec_version": {
-				Type:        schema.TypeInt,
+			"spec_version": schema.Int64Attribute{
 				Description: "Rule evaluation specification version in the case of breaking changes.",
-				Default:     1,
+				Computed:    true,
 				Optional:    true,
+				PlanModifiers: []planmodifier.Int64{
+					Int64DefaultValue(types.Int64Value(1)),
+				},
```

Comment:
In your testing did you find that `spec_version` is computed? I think there is only one spec version ("1") at the moment, but I could be wrong about that.

Or does `Computed` just mean in the case of no value we will compute one to add to the state?

Reply:
Yes, in the framework, it means the provider can provide a value if it isn't set, so "computed" must be true to use a default value in the framework. This is different from the previous SDK, but allows the acceptance tests to detect certain classes of provider bugs that couldn't be before.

>Setting the Computed property to true indicates that the attribute can be set by the provider. Any attribute that does not have Computed set to true cannot be influenced by the provider; it can only be changed via the config or via state refreshing.

https://developer.hashicorp.com/terraform/plugin/framework/handling-data/schemas#computed

Reply:
Thank you for the explanation. The default seems good to me!